fuzzers/ensemble/fuzzer.py

```diff
@@ -0,0 +1,145 @@
+    os.environ['AFL_I_DONT_CARE_ABOUT_MISSING_CRASHES'] = '1'
+
+    # AFL needs at least one non-empty seed to start.
+    if len(os.listdir(input_corpus)) == 0:
+        with open(os.path.join(input_corpus, 'default_seed'),
+                  'w') as file_handle:
+            file_handle.write('hi')
+
+
+# pylint: disable=too-many-arguments
+def run_afl_fuzz(fuzzer_name,
+                 input_corpus,
+                 output_corpus,
+                 target_binary,
+                 timeout,
+                 additional_flags=None,
+                 hide_output=False):
+    """Run afl-fuzz."""
+    # Spawn the afl fuzzing process.
+    # FIXME: Currently AFL will exit if it encounters a crashing input in seed
+    # corpus (usually timeouts). Add a way to skip/delete such inputs and
+    # re-run AFL. This currently happens with a seed in wpantund benchmark.
+    print('[run_fuzzer] Running target with afl-fuzz')
+    command = [
+        os.path.join('.', fuzzer_name),
+        '-i',
+        input_corpus,
+        '-o',
+        output_corpus,
+        # Use deterministic mode as it does best when we don't have
+        # seeds which is often the case.
+        '-d',
+        # Use no memory limit as ASAN doesn't play nicely with one.
+        '-m',
+        'none'
+    ]
+    if additional_flags:
+        command.extend(additional_flags)
+    command += [
+        '--',
+        target_binary,
+        # Pass INT_MAX to afl the maximize the number of persistent loops it
+        # performs.
+        '2147483647'
+    ]
+    output_stream = subprocess.DEVNULL if hide_output else None
+    subprocess.call(command,
+                    stdout=output_stream,
+                    stderr=output_stream,
+                    timeout=timeout)
+
+
+def fuzz(input_corpus, output_corpus, target_binary):
+    """Run afl-fuzz on target."""
+    prepare_fuzz_environment(input_corpus)
+    fuzzer_queue = collections.deque(
+        ['fair', 'mopt', 'afl', 'aflpp', 'aflsmart'], maxlen=5)
+
+    tmp_dir = tempfile.TemporaryDirectory()
+    next_fuzzer_input_corpus = os.path.join(str(tmp_dir), 'input_corpus')
+    shutil.copytree(input_corpus, next_fuzzer_input_corpus)
+    while True:
+        fuzzer_name = fuzzer_queue[0]
+        try:
+            fuzzer_output_dir = os.path.join(output_corpus, fuzzer_name)
+            run_afl_fuzz(fuzzer_name, next_fuzzer_input_corpus,
+                         fuzzer_output_dir, target_binary, SECONDS_PER_FUZZER)
+        except subprocess.TimeoutExpired:
+            tmp_dir.cleanup()
+            fuzzer_queue.rotate(1)
+            next_fuzzer_name = fuzzer_name
+            print('Switching to fuzzer to {0}.'.format(next_fuzzer_name))
+            tmp_dir = tempfile.TemporaryDirectory()
+            next_fuzzer_input_corpus = os.path.join(str(tmp_dir),
+                                                    'input_corpus')
+            shutil.copytree(os.path.join(fuzzer_output_dir, 'queue'),
```

Comment:
What happens the second time a fuzzer is run. Will copytree on the same next_fuzzer_input_corpus twice work?